File: apple/HDHROpenKit/Sources/HDHROpenKit/Playback/CaptionController.swift
import Foundation

@MainActor
public final class CaptionController: ObservableObject {
    @Published public private(set) var activeCueText: String?
    @Published public var isEnabled = false {
        didSet {
            if !isEnabled {
                activeCueText = nil
            }
        }
    }

    var cues: [CaptionCue] = []
    private var baseOffset = 0.0

    public init() {}

    public func setCues(_ cues: [CaptionCue], baseOffset: Double = 0.0) {
        self.cues = cues
        self.baseOffset = baseOffset
    }

    public func reset() {
        cues = []
        baseOffset = 0.0
        activeCueText = nil
    }

    public func updatePlaybackTime(_ currentTime: Double) {
        guard isEnabled, !cues.isEmpty else {
            activeCueText = nil
            return
        }

        let absoluteTime = currentTime + baseOffset
        if let matchingCue = cues.first(where: { $0.contains(time: absoluteTime) }) {
            activeCueText = matchingCue.text
        } else {
            activeCueText = nil
        }
    }
}

File: android/core/src/main/kotlin/org/hdhropen/kit/playback/CaptionController.kt
package org.hdhropen.kit.playback

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class CaptionController {
    private val _cues = MutableStateFlow<List<CaptionCue>>(emptyList())
    val cues: StateFlow<List<CaptionCue>> = _cues.asStateFlow()

    private val _isEnabled = MutableStateFlow(false)
    val isEnabled: StateFlow<Boolean> = _isEnabled.asStateFlow()

    private val _activeCueText = MutableStateFlow<String?>(null)
    val activeCueText: StateFlow<String?> = _activeCueText.asStateFlow()

    fun setCues(newCues: List<CaptionCue>) {
        _cues.value = newCues
    }

    fun toggleEnabled() {
        _isEnabled.value = !_isEnabled.value
        if (!_isEnabled.value) {
            _activeCueText.value = null
        }
    }

    fun setEnabled(enabled: Boolean) {
        _isEnabled.value = enabled
        if (!enabled) {
            _activeCueText.value = null
        }
    }

    fun updatePlaybackTime(currentTime: Double) {
        if (!_isEnabled.value) {
            _activeCueText.value = null
            return
        }

        val matchingCue = _cues.value.firstOrNull { it.contains(currentTime) }
        _activeCueText.value = matchingCue?.text
    }

    fun reset() {
        _cues.value = emptyList()
        _activeCueText.value = null
    }
}
